File: python3/second_minimum_node_in_a_binary_tree.py
from .model import TreeNode

# Space : O(n)
# Time  : O(n)


class Solution:
    def findSecondMinimumValue(self, root: TreeNode) -> int:
        s = [root]
        ans = -1
        minv = None

        while s:
            node = s.pop(0)

            if not minv:
                minv = node.val
            elif (ans == -1 or node.val < ans) and node.val != minv:
                if node.val < minv:
                    minv, ans = node.val, minv
                else:
                    ans = node.val

            if node.left:
                s.append(node.left)
            if node.right:
                s.append(node.right)

        return ans

File: golang/second_minimum_node_in_a_binary_tree.go
package main

import "golang/model"

// Space : O(n)
// Time  : O(n)

func findSecondMinimumValue(root *model.TreeNode) int {
    s := []*model.TreeNode{root}
    ans := -1
    minv := root.Val
    
    for len(s) > 0 {
        node := s[0]
        s = s[1:]
        
        if (ans == -1 || node.Val < ans) && node.Val != minv {
            if node.Val < minv {
                minv, ans = node.Val, minv
            } else {
                ans = node.Val
            }
        } 
        
        if node.Left != nil {
            s = append(s, node.Left)
        }
        if node.Right != nil {
            s = append(s, node.Right)
        }
    }
    
    return ans
}